File: 28.py

def create_odd_spiral(spiral):
    spiral_dim = len(spiral[0])
    last_number = spiral[0][-1]
    last_number += 1
    for i in range(spiral_dim):
        spiral[0 + i] = spiral[0 + i] + [last_number]
        last_number += 1
    spiral.append([])
    for i in range(spiral_dim + 2):
        spiral[-1] = [last_number] + spiral[-1]
        last_number += 1
    for i in range(spiral_dim):
        spiral[-2 - i] = [last_number] + spiral[-2 - i]
        last_number += 1
    spiral = [[]] + spiral
    for i in range(spiral_dim + 2):
        spiral[0] = spiral[0] + [last_number]
        last_number += 1

    return spiral

spiral = [[1]]
for i in range(1001 // 2):
    spiral = create_odd_spiral(spiral)

s = 0
s += sum(spiral[i][i] for i in range(len(spiral)))
s += sum(spiral[i][-1-i] for i in range(len(spiral)))
s -= 1

print(s)

File: cpp/28.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> create_odd_spiral(vector<vector<int>> spiral){

    int spiral_dim = spiral[0].size();
    int last_number = spiral[0].back();
    vector<int> tmp {};
    last_number ++;

    for (int i = 0; i < spiral_dim; i++)
    {
        tmp = spiral[i];
        tmp.push_back(last_number);
        spiral[i] = tmp;
        last_number++;
    }

    tmp = {};
    spiral.push_back(tmp);
    for (int i = 0; i < spiral_dim + 2; i++)
    {
        tmp = {last_number};
        tmp.insert(tmp.end(),spiral.back().begin(),spiral.back().end());
        spiral.pop_back();
        spiral.push_back(tmp);
        last_number++;
    }

    int ind {};
    for (int i = 0; i < spiral_dim; i++)
    {
        ind = spiral.size() - 2 - i;
        tmp = {last_number};
        tmp.insert(tmp.end(),spiral[ind].begin(),spiral[ind].end());
        spiral[ind] = tmp;
        last_number++;
    }
    spiral.insert(spiral.begin(), {{}});

    for (int i = 0; i < spiral_dim + 2; i++)
    {
        tmp = spiral[0];
        tmp.push_back(last_number);
        spiral[0] = tmp;
        last_number++;
    }
    
    return spiral;
}


int main() {
    int sum {0};
    vector<vector<int>> spiral{{1}};

    for (int i = 0; i < 500; i++){
        spiral = create_odd_spiral(spiral);
    }

    for (int i = 0; i < spiral.size(); i++)
        sum += spiral[i][i];
    for (int i = 0; i < spiral.size(); i++)
        sum += spiral[i][spiral.size() -1 -i];

    
    cout << sum -1 << endl;
    return 0;
}